core/src/main/scala/quasar/datasource/url/Http4sUrlDatasourceModule.scala

```diff
@@ -233,11 +221,107 @@ class Http4sUrlDatasourceModule(clientBuilder: Http4sClientBuilder)
+
+                            Resource.liftF(makeSub) flatMap {
+                              case Right(reqUri) =>
+                                makeResult[F](client, path, reqUri, urlConfig.headers, urlConfig.format)
+                              case Left(error) =>
+                                Resource.suspend(req.flatMap(r =>
+                                  MonadResourceErr[F].raiseError[Resource[F, InternalResult[F]]](
+                                    ResourceError.ConnectionFailed(
+                                      r.req.uri.resourcePath,
+                                      Some("Substituting initial request failed: " + error.msg),
+                                      None))))
+                            }
+                          }
+
+                          val templatedReq: F[ApiRequest[TemplatedUri]] =
+                            substitutions.map(vs => Multistep.substituteUUIDs(api.map(_._2), vs))
+
+                          branching match {
+                            case None =>
+                              makeSubstitutions(templatedReq)
+
+                            case Some(pc) =>
+                              val storedResult: Resource[F, InternalResult[F]] =
+                                lookup(pc.parentId, acc)
+
+                              val back: Resource[F, InternalResult[F]] = storedResult flatMap { result =>
+                                // extract the target values
+                                val values: Stream[F, String] =
+                                  Multistep.extractFromResult[F](result, path, pc.valueConfig)
+
+                                // group the extracted values
+                                val gathered: Stream[F, String] = pc match {
+                                  case ParentConfig.Zome(_, _, infoConfig, size) =>
+                                    Multistep.gatherSome[F](values, infoConfig, size)
+                                  case ParentConfig.One(_, _) =>
+                                    values
+                                }
+
+                                // compute a result per group
+                                val results: Stream[F, InternalResult[F]] =
+                                  gathered flatMap { group =>
+                                    val req: F[ApiRequest[TemplatedUri]] =
+                                      templatedReq.map(r =>
+                                        Multistep.substituteUUIDs(r, Map(pc.parentId -> group)))
+
+                                    Stream.resource(makeSubstitutions(req))
+                                  }
+
+                                // combine the results into a single result
+                                results.compile.resource.fold(
+                                  Multistep.emptyResult[F](urlConfig.format))(
+                                  Multistep.combineResults[F](_, _, path))
```

Comment:
Instead of compiling the stream here, can we be lazier?
```suggestion
                                InternalTypedResult(
                                  results.flatMap(Multistep.toTyped(_, path).data),
                                  urlConfig.format)
```

Reply:
It'd have to be
```scala
                                Resource.liftF(ConcurrentEffect[F].delay(
                                  InternalTypedResult(
                                    results.flatMap(Multistep.toTyped(_, path).data),
                                    urlConfig.format)))
```

Reply:
I guess that's okay?

Reply:
I think we can change the `flatMap` on L293 to `map` and avoid the extra lifting.

Reply:
Yes, that works!